trees/cart_regressor.go

```diff
@@ -0,0 +1,392 @@
+	}
+	if tree.Right == nil {
+		returnString += spacing + "---> False" + "\n"
+		returnString += "  " + spacing + "PREDICT    "
+		returnString += fmt.Sprintf("%.3f", tree.RightPred) + "\n"
+	}
+
+	if tree.Left != nil {
+		returnString += spacing + "---> True" + "\n"
+		returnString += regressorPrintTreeFromNode(*tree.Left, spacing+"  ")
+	}
+
+	if tree.Right != nil {
+		returnString += spacing + "---> False" + "\n"
+		returnString += regressorPrintTreeFromNode(*tree.Right, spacing+"  ")
+	}
+
+	return returnString
+}
+
+// Predict a single data point by navigating to rootNodes.
+// Uses a recursive logic
+func regressorPredictSingle(tree regressorNode, instance []float64) float64 {
+	if instance[tree.Feature] < tree.Threshold {
+		if tree.Left == nil {
+			return tree.LeftPred
+		} else {
+			return regressorPredictSingle(*tree.Left, instance)
+		}
+	} else {
+		if tree.Right == nil {
+			return tree.RightPred
+		} else {
+			return regressorPredictSingle(*tree.Right, instance)
+		}
+	}
+}
+
+// Predict method for multiple data points.
+// First converts input data into usable format, and then calls regressorPredictFromNode
+func (tree *CARTDecisionTreeRegressor) Predict(X_test base.FixedDataGrid) []float64 {
+	root := *tree.RootNode
+	test := regressorConvertInstancesToProblemVec(X_test)
+	return regressorPredictFromNode(root, test)
+}
+
+// Use tree's root node to print out entire tree.
+// Iterates over all data points and calls regressorPredictSingle to predict individual datapoints.
+func regressorPredictFromNode(tree regressorNode, test [][]float64) []float64 {
+	var preds []float64
+	for i := range test {
+		i_pred := regressorPredictSingle(tree, test[i])
+		preds = append(preds, i_pred)
+	}
+	return preds
+}
+
+// Helper function to convert base.FixedDataGrid into required format. Called in Fit, Predict
+func regressorConvertInstancesToProblemVec(X base.FixedDataGrid) [][]float64 {
+	// Allocate problem array
+	_, rows := X.Size()
+	problemVec := make([][]float64, rows)
+
+	// Retrieve numeric non-class Attributes
+	numericAttrs := base.NonClassFloatAttributes(X)
+	numericAttrSpecs := base.ResolveAttributes(X, numericAttrs)
+
+	// Convert each row
+	X.MapOverRows(numericAttrSpecs, func(row [][]byte, rowNo int) (bool, error) {
+		// Allocate a new row
+		probRow := make([]float64, len(numericAttrSpecs))
+		// Read out the row
+		for i, _ := range numericAttrSpecs {
+			probRow[i] = base.UnpackBytesToFloat(row[i])
+		}
+		// Add the row
+		problemVec[rowNo] = probRow
+		return true, nil
+	})
+	return problemVec
+}
+
+// Helper function to convert base.FixedDataGrid into required format. Called in Fit, Predict
+func regressorConvertInstancesToLabelVec(X base.FixedDataGrid) []float64 {
+	// Get the class Attributes
+	classAttrs := X.AllClassAttributes()
+	// Only support 1 class Attribute
+	if len(classAttrs) != 1 {
+		panic(fmt.Sprintf("%d ClassAttributes (1 expected)", len(classAttrs)))
```

Comment:
Never panic. All methods should return error.
Whenever there is error use fmt.Errorf("") and return.
When there is no error return nil.
Check id3.go for reference.

Reply:
Everytime you call a method inside a method and the called method returns error, the caller should check for the error and return error if it receives an error. 
Example code in id3.go
serializer, err := base.CreateSerializedClassifierStub(filePath, metadata)
if err != nil {
	return  err
}